### PR TITLE
Add launch_stack launchConfiguration type to ASG

### PR DESCRIPTION
**Pull request by jasondunsmore**

### CODE REVIEW

File: pyrax/autoscale.py [hunk 1]
```diff
@@ -388,28 +411,31 @@ def get_launch_config(self, scaling_group):
...
+            key_map = {
+                "OS-DCF:diskConfig": "disk_config",
+                "flavorRef": "flavor",
+                "imageRef": "image",
+            }
+            for key, value in args.get("server", {}).items():
+                norm_key = key_map.get(key, key)
+                ret[norm_key] = value
+        elif ret['type'] == 'launch_stack':
+            args.get("stack", {})
```
Comment (rs-randallburt) [inline pyrax/autoscale.py]: Shouldn't this arg be "template"? Is it called "stack" in the autoscale api!?

File: pyrax/autoscale.py [hunk 2]
```diff
@@ -779,35 +826,65 @@ def _encode_personality(self, personality):
...
+            disk_config = kwargs.get('disk_config', None)
+            metadata = kwargs.get('metadata', {})
+            personality = kwargs.get('personality')
+            networks = kwargs.get('networks')
+            load_balancers = kwargs.get('load_balancers')
+            key_name = kwargs.get('key_name')
+            config_drive = kwargs.get('config_drive', False)
+            user_data = kwargs.get('user_data')
+
+            return self._create_server_launch_config_body(
```
Comment (rs-randallburt) [inline pyrax/autoscale.py]: Seems this would be cleaner if there were just two separate methods that create the appropriate config based on type.

File: pyrax/autoscale.py (hunk 1 above)
Comment (dragorosson) [inline pyrax/autoscale.py]: No! :) and yes.

Comment (rs-randallburt) [inline pyrax/autoscale.py]: Oh, right.

**Comment by rs-randallburt** [comment]
:-1:  I think this needs a little more refactoring. Its a little hard to follow.

File: pyrax/autoscale.py [hunk 3]
```diff
@@ -417,19 +443,14 @@ def replace_launch_config(self, scaling_group, launch_config_type,
...
+        body = self._create_launch_config_body(launch_config_type,
+                                               **kwargs)
         resp, resp_body = self.api.method_put(uri, body=body)
 
 
-    def update_launch_config(self, scaling_group, server_name=None, image=None,
-            flavor=None, disk_config=None, metadata=None, personality=None,
-            networks=None, load_balancers=None, key_name=None, config_drive=False,
-            user_data=None):
+    def update_launch_config(self, scaling_group, launch_config_type,
```
Comment (dragorosson) [inline pyrax/autoscale.py]: I like the new signature of this method. However, I would make this method simply delegate to either `_update_server_launch_config` or `_update_stack_launch_config` depending on `launch_config_type`, where `_update_server_launch_config` would be the old (renamed) method (with the explicit keywords still in the signature).

Same for `_create_launch_config_body` and the other one/two similar methods. This is basically the strategy I used for implementing launch_stack in otter itself.

File: pyrax/autoscale.py [hunk 4]
```diff
@@ -438,48 +459,74 @@ def update_launch_config(self, scaling_group, server_name=None, image=None,
...
-        body = {"type": "launch_server",
-                "args": {
-                    "server": {
-                        "name": server_name or srv_args.get("name"),
-                        "imageRef": image or srv_args.get("imageRef"),
-                        "flavorRef": flav,
-                        "OS-DCF:diskConfig": dconf,
-                        "networks": networks or srv_args.get("networks"),
+        if launch_config_type == 'launch_server':
+            server_name = kwargs.get('server_name')
```
Comment (dragorosson) [inline pyrax/autoscale.py]: Referring to above, I think all of this indentation would be made unnecessary too.

**Comment by dragorosson** [comment]
:-1: Similar to what @rs-randallburt said. I didn't look too closely at the new code because I feel like it's sprinkled among a lot of line changes that are avoidable.

**Comment by jasondunsmore** [comment]
Thanks.  I've done some refactoring.  Please look again

File: pyrax/autoscale.py [hunk 5]
```diff
@@ -115,11 +113,36 @@ def update_launch_config(self, server_name=None, image=None, flavor=None,
         If you want to add to it, you either need to pass the complete dict of
         metadata, or call the update_launch_metadata() method.
         """
-        return self.manager.update_launch_config(self, server_name=server_name,
-                image=image, flavor=flavor, disk_config=disk_config,
-                metadata=metadata, personality=personality, networks=networks,
+        if launch_config_type == 'launch_server':
```
Comment (rs-randallburt) [inline pyrax/autoscale.py]: This seems like a good candidate to separate into two functions for clarity.

**Comment by rs-randallburt** [comment]
I think this still needs more work. At its core, it seems like a couple of methods for creating the appropriate dictionaries based on the type of launch configuration and then using those results in other places. Liberal use of `**kwargs` also seems like it would simplify a  lot of this code.

File: pyrax/autoscale.py (hunk 5 above)
Comment (dragorosson) [inline pyrax/autoscale.py]: +1

File: pyrax/autoscale.py [hunk 6]
```diff
@@ -477,9 +492,70 @@ def update_launch_config(self, scaling_group, server_name=None, image=None,
...
+            flavor = kwargs.get('flavor')
+            disk_config = kwargs.get('disk_config', None)
+            metadata = kwargs.get('metadata', {})
+            personality = kwargs.get('personality')
+            networks = kwargs.get('networks')
+            load_balancers = kwargs.get('load_balancers')
+            key_name = kwargs.get('key_name')
+            config_drive = kwargs.get('config_drive', False)
+            user_data = kwargs.get('user_data')
+            body = self._update_server_launch_config_body(
```
Comment (dragorosson) [inline pyrax/autoscale.py]: This is going in the right direction, but this section should really be

```
if launch_config_type == 'launch_server':
    body = self._update_server_launch_config_body(**kwargs)
```

Comment (jasondunsmore) [inline pyrax/autoscale.py]: But didn't you previously suggest that I do it this way?
"where _update_server_launch_config would be the old (renamed) method with the explicit keywords still in the signature"

**Comment by jasondunsmore** [comment]
Thanks.  I've done some more refactoring.  Please see the latest change

File: pyrax/autoscale.py (hunk 6 above)
Comment (dragorosson) [inline pyrax/autoscale.py]: Sorry for the confusion, I meant in the signature of the method (where it's defined), not when calling it.

File: pyrax/autoscale.py [hunk 7]
```diff
@@ -417,30 +414,26 @@ def replace_launch_config(self, scaling_group, launch_config_type,
...
+        networks = kwargs.get('networks')
+        load_balancers = kwargs.get('load_balancers')
+        key_name = kwargs.get('key_name')
+        config_drive = kwargs.get('config_drive', False)
+        user_data = kwargs.get('user_data')
 
-    def update_launch_config(self, scaling_group, server_name=None, image=None,
-            flavor=None, disk_config=None, metadata=None, personality=None,
-            networks=None, load_balancers=None, key_name=None, config_drive=False,
-            user_data=None):
```
Comment (dragorosson) [inline pyrax/autoscale.py]: I think all of the signatures that got their keyword arguments replaced with **kwargs can be left alone instead of using the `parameter = kwargs.get('parameter')` pattern. It achieves the same effect, right?

File: pyrax/autoscale.py [hunk 8]
```diff
@@ -477,9 +470,50 @@ def update_launch_config(self, scaling_group, server_name=None, image=None,
...
+        return body
+
+    def _update_stack_launch_config_body(self, **kwargs):
+        template = kwargs.get('template')
+        template_url = kwargs.get('template_url')
+        disable_rollback = kwargs.get('disable_rollback')
+        environment = kwargs.get('environment')
+        files = kwargs.get('files')
+        parameters = kwargs.get('parameters')
+        timeout_mins = kwargs.get('timeout_mins')
```
Comment (dragorosson) [inline pyrax/autoscale.py]: Similar to above, put these in the signature instead of using `kwargs`.

File: pyrax/autoscale.py [hunk 9]
```diff
@@ -823,11 +860,40 @@ def _create_group_config_body(self, name, cooldown, min_entities,
                 }
         return body
 
-
-    def _create_launch_config_body(self, launch_config_type,
-            server_name, image, flavor, disk_config=None, metadata=None,
-            personality=None, networks=None, load_balancers=None,
-            key_name=None, config_drive=False, user_data=None):
```
Comment (dragorosson) [inline pyrax/autoscale.py]: Ditto

File: pyrax/autoscale.py [hunk 10]
```diff
@@ -823,11 +860,40 @@ def _create_group_config_body(self, name, cooldown, min_entities,
...
+    def _create_stack_launch_config_body(self, **kwargs):
+        template = kwargs.get('template')
+        template_url = kwargs.get('template_url')
+        disable_rollback = kwargs.get('disable_rollback')
+        environment = kwargs.get('environment')
```
Comment (dragorosson) [inline pyrax/autoscale.py]: Ditto

File: pyrax/autoscale.py [hunk 11]
```diff
@@ -823,11 +860,40 @@ def _create_group_config_body(self, name, cooldown, min_entities,
...
+        image = kwargs.get('image')
+        flavor = kwargs.get('flavor')
+        disk_config = kwargs.get('disk_config', None)
+        metadata = kwargs.get('metadata', {})
+        personality = kwargs.get('personality')
+        networks = kwargs.get('networks')
+        load_balancers = kwargs.get('load_balancers')
+        key_name = kwargs.get('key_name')
+        config_drive = kwargs.get('config_drive', False)
+        user_data = kwargs.get('user_data')
```
Comment (dragorosson) [inline pyrax/autoscale.py]: Ditto (though this belongs to the removed code that was dittoed above)

**Comment by dragorosson** [comment]
Looking good apart from the kwargs stuff, which is getting there.

File: pyrax/autoscale.py (hunk 1 above)
Comment (dragorosson) [inline pyrax/autoscale.py]: This line doesn't actually do anything...

File: pyrax/autoscale.py [hunk 12]
```diff
@@ -388,28 +382,31 @@ def get_launch_config(self, scaling_group):
...
+                "OS-DCF:diskConfig": "disk_config",
+                "flavorRef": "flavor",
+                "imageRef": "image",
+            }
+            for key, value in args.get("server", {}).items():
+                norm_key = key_map.get(key, key)
+                ret[norm_key] = value
+        elif ret['type'] == 'launch_stack':
+            args.get("stack", {})
+            for key, value in args.get("stack", {}).items():
```
Comment (dragorosson) [inline pyrax/autoscale.py]: If this is supposed to return the launch config, I'd think that something really simple like `ret['args'] = args` or `ret['args'] = {'stack': args.get('stack', {})}` would be sufficient. Since there aren't any keys that have to be normalized/remapped, the launch stack code is just returning a shallow copy of the `stack` mapping.

... but puts it in the top level of the launch config. Correct me if I'm wrong, but doesn't this return something like
```
{'type': 'launch_stack',
 'template': ...,
 'parameters': ...,
 ...}
```
even though launch configs are supposed to be of the form `{'type': ..., 'args': {...}}`. I see that the original code does it this strange way. I think it's a bug.

File: pyrax/autoscale.py [hunk 13]
```diff
@@ -477,9 +470,50 @@ def update_launch_config(self, scaling_group, server_name=None, image=None,
...
+            'template_url': template_url,
+            'disable_rollback': disable_rollback,
+            'environment': environment,
+            'files': files,
+            'parameters': parameters,
+            'timeout_mins': timeout_mins
+        }
+        st_args = {
+            k: st_args[k] for k in st_args if st_args[k] is not None}
+        return {"type": 'launch_stack', "args": {"stack": st_args}}
```
Comment (dragorosson) [inline pyrax/autoscale.py]: Is this method exactly the same as `_create_stack_launch_config_body`? If so, I'd remove one and/or have one call the other.

File: pyrax/autoscale.py [hunk 14]
```diff
@@ -477,9 +470,50 @@ def update_launch_config(self, scaling_group, server_name=None, image=None,
         key_name = key_name or srv_args.get("key_name")
         if key_name:
             bas["key_name"] = key_name
-        resp, resp_body = self.api.method_put(uri, body=body)
-        return None
+        return body
+
+    def _update_stack_launch_config_body(self, **kwargs):
```
Comment (dragorosson) [inline pyrax/autoscale.py]: It looks like this method is exactly the same as `_create_stack_launch_config_body`.

File: pyrax/autoscale.py [hunk 15]
```diff
@@ -477,9 +460,37 @@ def update_launch_config(self, scaling_group, server_name=None, image=None,
         key_name = key_name or srv_args.get("key_name")
         if key_name:
             bas["key_name"] = key_name
-        resp, resp_body = self.api.method_put(uri, body=body)
-        return None
+        return body
 
+    def _update_stack_launch_config_body(
+            self, template=None, template_url=None, disable_rollback=True,
```
Comment (dragorosson) [inline pyrax/autoscale.py]: I think this can delegate through **kwargs instead of the signature and the call to `_create...` specifying all of the keyword arguments.

File: pyrax/autoscale.py [hunk 16]
```diff
@@ -477,9 +460,37 @@ def update_launch_config(self, scaling_group, server_name=None, image=None,
...
+        One or more of the available attributes can be specified.
+
+        NOTE: if you specify metadata, it will *replace* any existing metadata.
+        If you want to add to it, you either need to pass the complete dict of
+        metadata, or call the update_launch_metadata() method.
+        """
+        uri = "/%s/%s/launch" % (self.uri_base, scaling_group.id)
+
+        if launch_config_type == 'launch_server':
+            body = self._update_server(
```
Comment (dragorosson) [inline pyrax/autoscale.py]: Incomplete method name?

File: pyrax/autoscale.py [hunk 17]
```diff
@@ -779,35 +784,38 @@ def _encode_personality(self, personality):
...
-            launch_config_type, server_name, image, flavor, disk_config=None,
-            metadata=None, personality=None, networks=None,
-            load_balancers=None, scaling_policies=None, group_metadata=None,
-            key_name=None, config_drive=False, user_data=None):
+            launch_config_type, **kwargs):
         """
         Used to create the dict required to create any of the following:
             A Scaling Group
         """
-        if metadata is None:
```
Comment (dragorosson) [inline pyrax/autoscale.py]: I think this disappeared without being moved to _create_server_launch_config_body

File: pyrax/autoscale.py [hunk 18]
```diff
@@ -115,11 +113,7 @@ def update_launch_config(self, server_name=None, image=None, flavor=None,
         If you want to add to it, you either need to pass the complete dict of
         metadata, or call the update_launch_metadata() method.
         """
-        return self.manager.update_launch_config(self, server_name=server_name,
-                image=image, flavor=flavor, disk_config=disk_config,
-                metadata=metadata, personality=personality, networks=networks,
-                load_balancers=load_balancers, key_name=key_name,
-                config_drive=config_drive, user_data=user_data)
+        return self.manager.update_launch_config(**kwargs)
```
Comment (dragorosson) [inline pyrax/autoscale.py]: launch_config_type got missed here

**Comment by dragorosson** [comment]
Looking good. From only looking through the code, everything looks right, and I'm sure most of the remaining small issues will be flushed out through unit testing.

**Comment by jasondunsmore** [comment]
Uploaded unit tests.  Please check again.
